models/VQVAE.py

```diff
@@ -0,0 +1,280 @@
+        super(Decoder, self).__init__()
+
+        # Initializations
+        self.upsampling_layers = nn.ModuleList()
+        self.residual_stack = ResidualStack(num_hiddens, num_residual_layers, num_residual_hiddens)
+
+        # Layer Definitions
+        self.initial_conv = nn.Conv2d(in_channels=embedding_dim, out_channels=num_hiddens, kernel_size=3, padding=1)
+
+        for i in range(num_upsampling_layers):
+            if i < num_upsampling_layers - 2:
+                in_channels, out_channels = num_hiddens, num_hiddens
+            elif i == num_upsampling_layers - 2:
+                in_channels, out_channels = num_hiddens, num_hiddens // 2
+            else:
+                in_channels, out_channels = num_hiddens // 2, 3
+
+            self.upsampling_layers.append(
+                nn.Sequential(
+                    nn.ConvTranspose2d(in_channels=in_channels, out_channels=out_channels, kernel_size=4, stride=2, padding=1),
+                    nn.ReLU() if i < num_upsampling_layers - 1 else nn.Identity(),
+                )
+            )
+
+    def forward(self, x):
+        # Forward Pass
+        x = self.initial_conv(x)
+        x = self.residual_stack(x)
+
+        for layer in self.upsampling_layers:
+            x = layer(x)
+
+        return x
+
+
+class VQVAE(nn.Module):
+    def __init__(self, cfg, device):
+        super().__init__()
+
+        in_channels = cfg.in_channels
+        num_hiddens = cfg.num_hiddens
+        num_downsampling_layers = cfg.num_downsampling_layers
+        num_residual_layers = cfg.num_residual_layers
+        num_residual_hiddens = cfg.num_residual_hiddens
+        embedding_dim = cfg.embedding_dim
+        num_embeddings = cfg.num_embeddings
+        use_ema = cfg.use_ema
+        decay = cfg.decay
+        epsilon = cfg.epsilon
+
+        self.device = torch.device("cuda" if device and torch.cuda.is_available() else "cpu")
+        print(self.device)
+
+        # Define layers in __init__
+        self.encoder_layer = self.build_encoder(
+            in_channels,
+            num_hiddens,
+            num_downsampling_layers,
+            num_residual_layers,
+            num_residual_hiddens,
+        )
+        self.pre_vq_conv_layer = nn.Conv2d(
+            in_channels=num_hiddens,
+            out_channels=embedding_dim,
+            kernel_size=1,
+        )
+        self.vector_quantizer = self.build_vector_quantizer(
+            embedding_dim,
+            num_embeddings,
+            use_ema,
+            decay,
+            epsilon,
+        )
+        self.decoder_layer = self.build_decoder(
+            embedding_dim,
+            num_hiddens,
+            num_downsampling_layers,
+            num_residual_layers,
+            num_residual_hiddens,
+        )
+
+    def build_encoder(self, in_channels, num_hiddens, num_downsampling_layers, num_residual_layers, num_residual_hiddens):
```

Comment:
```suggestion
    def build_encoder(
        self,
        in_channels: int,
        num_hiddens: int,
        num_downsampling_layers: int,
        num_residual_layers: int,
        num_residual_hiddens: int,
    ) -> Encoder:
```

Making reading the code as straightforward as possible is our priority here, and type hint (including return) and per-line arguments is extremely helpful in this (for me at least).

You can always ask chat gpt to write type hints for you :sweat_smile: